nemo/collections/nlp/metrics/qa_metrics.py

```diff
@@ -0,0 +1,199 @@
+        for output in outputs:
+            for i, key in enumerate(keys):
+                if isinstance(output[key], torch.Tensor):
+                    output_lists[i].extend(tensor2list(output[key]))
+                else:
+                    output_lists[i].extend(output[key])
+
+        return output_lists
+
+    @staticmethod
+    def get_exact_match_and_f1(examples, preds, question_id_filter=[]):
+        """
+        Returns a dictionary of question id: exact match/f1 score
+        Questions with ids *not* present in `question_id_filter` are excluded
+        """
+        exact_scores = {}
+        f1_scores = {}
+
+        for example in examples:
+            question_id = example.qas_id
+            if question_id not in question_id_filter:
+                continue
+
+            gold_answers = [answer["text"] for answer in example.answers if QAMetrics.normalize_answer(answer["text"])]
+
+            if not gold_answers:
+                # For unanswerable questions, only correct answer is empty string
+                gold_answers = [""]
+
+            pred = preds[question_id]
+            exact_scores[question_id] = max(QAMetrics.get_one_exact_match(pred, a) for a in gold_answers)
+            f1_scores[question_id] = max(QAMetrics.get_one_f1(pred, a) for a in gold_answers)
+
+        return exact_scores, f1_scores
+
+    @staticmethod
+    def make_eval_dict(exact_scores, f1_scores, prefix=""):
+        """ Returns dictionary with formatted evaluation scores """
+
+        total = len(exact_scores)
+        return collections.OrderedDict(
+            [
+                (f"{prefix}exact", (100.0 * sum(exact_scores.values()) / total) if total != 0 else 0.),
+                (f"{prefix}f1", (100.0 * sum(f1_scores.values()) / total) if total != 0 else 0.),
+                (f"{prefix}total", total),
+            ]
+        )
+
+    @staticmethod
+    def merge_eval_dicts(eval_dicts):
+        """
+        Combines multiple evaluation dict outputs into one dict
+        Ex: combines eval dicts for HasAns F1, NoAnsF1, and Total F1
+        """
+
+        merged_dict = collections.OrderedDict()
+        for eval_dict in eval_dicts:
+            for key in eval_dict:
+                merged_dict[key] = eval_dict[key]
+
+        return merged_dict
+
+    @staticmethod
+    def evaluate_predictions(examples, all_predictions):
+        """ 
+        Calculates exact match and f1 scores for all predictions, 
+            questions with answers, and no answer questions
+        """
+
+        qas_id_to_has_answer = {example.qas_id: bool(example.answers) for example in examples[: len(all_predictions)]}
+        has_answer_qids = [qas_id for qas_id, has_answer in qas_id_to_has_answer.items() if has_answer]
+        no_answer_qids = [qas_id for qas_id, has_answer in qas_id_to_has_answer.items() if not has_answer]
+
+        all_exact, all_f1 = QAMetrics.get_exact_match_and_f1(examples, all_predictions, list(qas_id_to_has_answer))
```

Comment:
nit: Looks like you're repeating yourself a few times. if it's fast to write, can you maybe write a for loop instead?